Multiple sites on the same hosts enhancements

mamedin opened the pull request:
This PR adds:

* Use a different revision directory for every site update.
* Add new tasks file for CLI commands.
* Add private symlink.
* Add downloads symlink.
* Ensure downloads and uploads symlink directories exist.

Connects to https://github.com/artefactual-labs/ansible-atom/issues/53 and #20

scollazo commented on defaults/main.yml:
```diff
@@ -36,11 +44,25 @@ atom_worker_service_name: "{{ atom_path | basename }}-worker"
+# Upgrade sql
+atom_upgrade_sql: "no"
+# Rebuild nested-set and generate slugs
+atom_fix_data: "no"
```
Duplicate

scollazo commented on defaults/main.yml:
```diff
@@ -50,9 +72,9 @@ atom_build_static_assets: "yes"
 # Themes (to build)
 atom_themes:
-  - path: "{{ atom_path }}/plugins/arDominionPlugin"
+  - path: "{{ atom_path }}/{{ atom_extra_path }}/plugins/arDominionPlugin"
```
Why are you changing almost all the instances of atom_path to add atom_extra_path, instead of updating atom_path once and for all?

mamedin replied:
Do you mean redefining `atom_path` with `set_fact`? 

There's a problem when using -e "@sites/whatever/site.yml" and the ansible variable precedence:

https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#variable-precedence-where-should-i-put-a-variable

The `atom_path` is order 21 in this case and we are trying to change with a task var (order 17).

I don't like using the new `atom_extra_path` variable either, but I haven't found an easier way to do it.

mamedin replied:
A second alternative is to use a new variable (i.e `atom_src_path`) and define it in role with `set_fact`, but I think it is clearer using the `atom_extra_path` variable.

hakamine replied:
I like @mamedin 's approach of using a separate variable `atom_extra_path` (I think redefining `atom_path` variable inside the role would be confusing and probably make debugging/troubleshooting harder)

hakamine commented on README.md:
```diff
@@ -101,3 +101,23 @@ as `atom-site` in the role (i.e., tasks required to deploy a site)
+├── atom-381f849b6ecd763dedf92a7bad43cc350a3c5439
+├── downloads
+├── private -> /usr/share/nginx/test5/src
+├── src -> /usr/share/nginx/test5/atom-381f849b6ecd763dedf92a7bad43cc350a3c5439
```
In this example, should `test5` be `atom` ? (e.g., , `src` linking to `/usr/share/nginx/atom/atom-381f849b6ecd763dedf92a7bad43cc350a3c5439`? ; or `usr/share/nginx/atom` replaced with `/usr/share/nginx/test5` ?)

mamedin replied:
Hi Hector,

I deployed this site with:

```
atom_path: /usr/share/nginx/test5
```

hakamine commented on tasks/revision-dir.yml:
```diff
@@ -0,0 +1,19 @@
+    update: no
+  register: "atom_revision_git_output"
+  when: "atom_revision_directory|bool and atom_revision_directory_latest_symlink_dir is defined"
+
```
Does this git task with `clone:no` and `update:no` work even when doing the first deployment (a new deployment)? (From ansible docs: "If clone: no, do not clone the repository even if it does not exist locally". "If update:no, do not retrieve new revisions from the origin repository")

hakamine commented on tasks/cli_tools.yml:
```diff
@@ -0,0 +1,46 @@
+  shell: "cd {{ item }} && make"
+  with_items: "{{ theme_makefiles.stdout_lines }}"
+  when: "atom_compile_all_themes|bool"
+
```
There are already tasks for building the AtoM themes in tasks/build.yml

hakamine commented on tasks/symlink-dirs.yml:
```diff
@@ -0,0 +1,79 @@
+    - "atom_revision_directory_latest_symlink_dir is defined"
+    - "atom_revision_directory|bool"
+    - "atom_private_symlink|bool"
+
```
what is the purpose of the "private" symlink? Does AtoM need it?

DavidHume-ArtefactualSystems replied:
As discussed, I believe yes ; And now that I see it in this context, makes sense to me rather than a post_task, but whatever works. ¯\_(ツ)_/¯

hakamine commented on defaults/main.yml:
```diff
@@ -15,6 +15,14 @@ atom_install_dependencies: "true"
+atom_private_symlink: "no"
+#Please, don't change atom_extra_path, it should be ""
+atom_extra_path: ""
 
```
I was wondering if there could be a way to avoid defining `atom_extra_path` here (because this is a variable created/modified inside the role, and it is not a variable whose value could be customized by the user)

mamedin commented:
Thanks very much for CR!

I have changed the branch to include your suggestions. Using the summary points added by @hakamine in last comment:

0. Fixed duplicated variables added to defaults/main.yml
1. Fixed example in README.md
2. `atom_extra_path` is now defined with `set_fact`. I added this task in `tasks/main.yml` because this variable is used by `drmc-mock` and `atom-devbox` tasks. The "always" tag has been added to this task and the `include: "revision-dir.yml"` one because it is used by almost all tags, the explicit list should be:

    - "atom-site"
    - "drmc-mock"
    - "atom-devbox"
    - "atom-basic"
    - "atom-flush"
    - "atom-plugins"
    - "atom-search"
    - "atom-cli"
    - "atom-build"
    - "atom-worker"
    - "atom-uploads"
    - "atom-downloads"

3. (No changes from last PR branch) The "Compile all themes" task is useful to avoid having to explicitly declare the themes you want to build in the `atom_themes` dictionary. Do you think it is a good idea to add and additional conditional to the existing "Build AtoM themes (DISTRO)" tasks (to avoid compile 2 times the themes when `atom_compile_all_themes|bool`)? For instance:

     - not atom_compile_all_themes|bool

4. I have not detected any issue when the repository has not been already cloned before.
5. Probably the "private symlink" task doesn't make sense in the AtoM role. We can remove the "private symlink" commit and create create the post_task directly from the AtoM playbook when needed.

hakamine commented:
@mamedin thank you for the fixes. I have a comment regarding the use of the "always" tag in role, the tasks will execute every time a playbook is run and there is no way to exclude them. For example, the following playbook: 
```
- hosts: somehost
  roles:
    - role: "nginx"
      tags: nginx
    - role: "atom"
      tags: atom
```
even if running the playbook with tag nginx (e.g. `ansible-playbook playbook.yml -t nginx`) it will execute the tasks in the atom role tagged with "always". It's probably safe if the tasks do not modify the target host (assigning vars, etc), but it could be kind of confusing.

DavidHume-ArtefactualSystems commented:
@mamedin @hakamine - I believe for _atom_revision_directory_ style deploys of AtoM the **private** symlink is required for digital object access control (copyright etc.), at least was when originally set up, I'd have to recall which Hosted clients definitely used it (Mills?) and per my using this method for SFU this year, should be required.

Or Miguel are you saying it would still be present if the boolean is set, just not going to do it in the role now (number 5. in your lists).

mamedin commented:
> @mamedin @hakamine - I believe for _atom_revision_directory_ style deploys of AtoM the **private** symlink is required for digital object access control (copyright etc.), at least was when originally set up, I'd have to recall which Hosted clients definitely used it (Mills?) and per my using this method for SFU this year, should be required.
> 
> Or Miguel are you saying it would still be present if the boolean is set, just not going to do it in the role now (number 5. in your lists).

mamedin commented:
Ups, wrong click!

mamedin commented:
> @mamedin thank you for the fixes. I have a comment regarding the use of the "always" tag in role, the tasks will execute every time a playbook is run and there is no way to exclude them. For example, the following playbook:
> 
> ```
> - hosts: somehost
>   roles:
>     - role: "nginx"
>       tags: nginx
>     - role: "atom"
>       tags: atom
> ```
> 
> even if running the playbook with tag nginx (e.g. `ansible-playbook playbook.yml -t nginx`) it will execute the tasks in the atom role tagged with "always". It's probably safe if the tasks do not modify the target host (assigning vars, etc), but it could be kind of confusing.

I'm going to change the always tag to the explicit list. Thanks Hector!

mamedin commented:
Hi!, branch updated again. Using again the summary points, these are the new changes:

2.- "always" tag changed on revision tasks, the tasks are using the big tag list instead.
3.- Build all themes tasks have been moved to the build.yml task file.
7.- New commit. The "deph" and "force" options for git clone are now configurable. Using old values as default options.

DavidHume-ArtefactualSystems commented on tasks/cli_tools.yml:
```diff
@@ -0,0 +1,37 @@
+---
+
+# Compile available themes
```
(Cosmetic) Having moved this section back to tasks/build.yml, should remove this comment line.

mamedin replied:
Good catch!

DavidHume-ArtefactualSystems commented on tasks/cli_tools.yml:
```diff
@@ -0,0 +1,37 @@
+    chdir: "{{ atom_path }}/{{ atom_extra_path }}"
+  ignore_errors: "yes" # segfault in php7, not sure yet why!
+  environment: "{{ atom_pool_php_envs }}"
+  when: "atom_populate_index|bool"
```
Yeah, more consistent than 'force_search_rebuild', that'll be another of those translated variables for the group_vars/ inventory file or the translation file, maybe.  :-)

mamedin replied:
@DavidHume-ArtefactualSystems thanks! I forgot to copy it from our all playbook. I'll update the branch with this change ;)

mamedin replied:
Branch updated!

DavidHume-ArtefactualSystems commented on tasks/revision-dir.yml:
```diff
@@ -0,0 +1,16 @@
+  register: "atom_revision_git_output"
+
+- name: "Print revisions"
+  debug: "msg='current={{ atom_revision_git_output.before }} latest={{ atom_revision_git_output.after }}'"
```
In my tests, even with a previous revision these two results would be the same - but I think that might be because of the other issues we encountered. Hence my wondering @mamedin if you have gotten this output with different results yet. That is, _atom_revision_git_output.before_ value via akin to what we used to keep in the **latest** file before updating, and _atom_revision_git_output.after_ being the true current (which I assume the above first task does more efficiently than shell command "git ls-remote {{ atom_repository_url }} {{ atom_repository_version }} | awk '{print $1}'"  did).

That said, I think as we discussed as long as **.after** is correct (so won't try to overwrite a previous or fallback revision directory) all is well...

DavidHume-ArtefactualSystems replied:
@hakamine - did you have a similar concern, I guess this relates to your refactor suggestion for _revision-dir.yml_ on August 10 above?

hakamine replied:
@DavidHume-ArtefactualSystems when I first saw the PR code I was concerned that maybe using git with "clone: no" and "update:no" would fail if there was no local repo. But re-checking the ansible docs, they provide a usage example of git with "clone: no" and "update:no", to "just get information about the repository whether or not it has already been cloned locally."  therefore I think it should be good regarding my original concern. However, the problem you encountered seems to be a different issue.

mamedin replied:
Probably related to:

https://github.com/artefactual-labs/ansible-atom/blob/master/tasks/basic.yml#L3-L9

Using depth=1 is known problem.

luisilara commented:
I tested the new changes  on a client upgrade and it worked properly
- The `revision-dir.yml` improves the  command shell: "git ls-remote {{ atom_repository_url }} {{ atom_repository_version }} | awk '{print $1}'"  where in some cases it creates the revision folder with the stdout array eg. `'atom-d6ffd39c150ebe74adb0bc9e7e8d84797ad8def2'$'\n''c017acb60e8dceeb6ec453e368381bf12bc940ad'/`
- The new variable `atom_git_pull_code_depth ` will produce a shallow clone, probably the default value could be set to 1 to avoid adding that variable to the playbook?

mamedin commented:
> I tested the new changes on a client upgrade and it worked properly
> 
> * The `revision-dir.yml` improves the command `shell: "git ls-remote {{ atom_repository_url }} {{ atom_repository_version }} | awk '{print $1}'"` where in some cases it created the revision folder with the stdout with two lines eg.` 'atom-d6ffd39c150ebe74adb0bc9e7e8d84797ad8def2'$'\n''c017acb60e8dceeb6ec453e368381bf12bc940ad'/`
> * The new variable atom_git_pull_code_depth will produce a shallow clone, probably the default value could be set to 1 to avoid adding that variable to the playbook?

@luisilara The `omit` variable is an ansible special variable:

https://docs.ansible.com/ansible/latest/user_guide/playbooks_filters.html#omitting-parameters

So the default value is `depth: null` when running the playbook with -vvv option (the same value that you can see when not defining `depth` option in role).

hakamine commented:
@mamedin I think the `accept_hostkey: "yes"` parameter may also be needed in the git task in `tasks/revision-dir.yml` (currently only the git task in `tasks/basic.yml` has it). I just did a test and failed there.

hakamine commented:
@mamedin @DavidHume-ArtefactualSystems What do you think about removing the `{{ atom_path }}/{{ atom_extra_path }}` prefix in the `atom_themes` path items in `defaults/main.yml` (i.e., making the paths in atom_themes relative to the AtoM base directory), and instead adding this prefix to the "Build selected AtoM themes" tasks in `tasks/build.yml`. I think it could make the atom_themes definitions in host_vars simpler.

mamedin commented:
> @mamedin @DavidHume-ArtefactualSystems What do you think about removing the `{{ atom_path }}/{{ atom_extra_path }}` prefix in the `atom_themes` path items in `defaults/main.yml` (i.e., making the paths in atom_themes relative to the AtoM base directory), and instead adding this prefix to the "Build selected AtoM themes" tasks in `tasks/build.yml`. I think it could make the atom_themes definitions in host_vars simpler.

Yes, it will be a nice change, but the problem is that we already have 25 config files with the `atom_themes` dict in `deploy-private` repo. 

We are going to change some other settings on upgrades, so if you agree I can change it.

mamedin commented:
I have upgraded again the branch with this changes:

* The line `# Compile available themes` in cli_tools.yml has been removed
* Added `accept_hostkey: "yes"` to the revision task.

DavidHume-ArtefactualSystems commented:
> @mamedin I think the `accept_hostkey: "yes"` parameter may also be needed in the git task in `tasks/revision-dir.yml` (currently only the git task in `tasks/basic.yml` has it). I just did a test and failed there.

Great catch @hakamine !

hakamine commented:
@mamedin I ran a test for an upgrade (deployed a branch, then did another deploy with a different code commit), and I think I  got the same problem that @DavidHume-ArtefactualSystems reported. The first time I tried the upgrade, the "before:" and "after:" commit hashes reported by the git task were the same (old commit). I had to run the playbook again so that the upgrade was completed, this time the git task reported "before:" and "after:" commit hashes were the same but the new commit (I am NOT using depth=1 BTW). So this meaning that in order to do an upgrade would need to run the playbook twice. Could it be a bug in the ansible git module? Could it be better to stick with the old way (using git ls-remote) ?

mamedin commented:
> @mamedin I ran a test for an upgrade (deployed a branch, then did another deploy with a different code commit; strictly speaking it wasn't a branch update but a branch change), and I think I got the same problem that @DavidHume-ArtefactualSystems reported. The first time I tried the upgrade, the "before:" and "after:" commit hashes reported by the git task were the same (old commit). I had to run the playbook again so that the upgrade was completed, this time the git task reported "before:" and "after:" commit hashes were the same but the new commit (I am NOT using depth=1 BTW). So this meaning that in order to do an upgrade would need to run the playbook twice. Could it be a bug in the ansible git module? Could it be better to stick with the old way (using git ls-remote) ?

Thanks! I could reproduce the issue when using different branches. I could fix it with git module but using 2 different tasks for getting the current and candidate revision. The `dest` option is optional when using `clone=yes`, so I tested without the `dest` option for the candidate tasks and it fixes the issue.

I prefer to use the git module because the `git ls-remote` tasks requires `remote: 127.0.0.1` and sometimes it fails with AWX or when running the ansible playbook in a different server/VM (ssh forwarding issues).

mamedin commented:
Thanks again for suggestions :)

Branch updated again. The `revision-dir.yml` file has changed:

* Use `candidate` instead of `latest` for the revision to be installed. (Cosmetic)
* Use separate tasks to get the `candidate` and `current` revisions.
